File: python/src/threads/thread/read.py
"""A branch's log as a reader of the public handle sees it."""

from threads.agents.store import Store, now_ms, open_store
from threads.log import BranchId, ParseError
from threads.result import Err, Ok
from threads.store import VerifiedLog


async def read_log(store: Store, branch: BranchId) -> Ok[VerifiedLog] | Err[ParseError]:
    """The branch's verified log. A branch that fails verification is log_corrupt."""
    read = await (await open_store(store)).read(branch, now_ms())
    return read if isinstance(read, Ok) else Err(reader_error(read.error))


def reader_error(error: ParseError) -> ParseError:
    """Any failure but a missing branch or an unsupported line is log_corrupt."""
    match error.code:
        case "branch_not_found":
            return ParseError("not_found", error.message)
        case "unsupported_format" | "unsupported_critical_event":
            return error
        case _:
            return ParseError("log_corrupt", f"{error.code}: {error.message}", error.seq)

File: typescript/packages/core/src/thread/read.ts
import type { BranchId } from "../log";
import { err, type Result } from "../result";
import type { LogStore } from "../store";
import type { LogError, VerifiedLog } from "../verify";

/** The read errors of spec/api.json: why a stored log can't be read. */
export type ReadErrorCode =
  | "log_corrupt"
  | "unsupported_format"
  | "unsupported_critical_event";

export type ReadError = LogError & { readonly code: ReadErrorCode };

export function readError(
  code: ReadErrorCode,
  message: string,
  seq?: number,
): ReadError {
  return seq === undefined ? { code, message } : { code, message, seq };
}

/** A branch's log as a reader sees it; a failure other than an unsupported line is log_corrupt. */
export async function readLog(
  log: LogStore,
  branchId: BranchId,
): Promise<Result<VerifiedLog, ReadError>> {
  const read = await log.read(branchId);
  if (read.ok) return read;
  const { code, message, seq } = read.error;
  const unsupported =
    code === "unsupported_format" || code === "unsupported_critical_event";
  return err(readError(unsupported ? code : "log_corrupt", message, seq));
}
